secure/key-manager/src/tests.rs

```diff
@@ -251,58 +208,296 @@ impl LibraInterface for TestLibraInterface {
+    }
+
+    fn take_all_transactions(&self) -> Vec<Transaction> {
+        self.queued_transactions.borrow_mut().drain(0..).collect()
+    }
+}
+
+#[derive(Clone)]
+/// A JsonRpcLibraInterface wrapper implementation that stores a list of transactions submitted via
+/// the API and exposes additional methods for testing (by implementing the LibraInterfaceWrapper
+/// trait). This is required to better test the JsonRpcLibraInterface implementation.
+struct JsonRpcLibraWrapper {
+    submitted_transactions: Arc<RefCell<Vec<Transaction>>>,
+    libra_interface: JsonRpcLibraInterface,
+}
+
+impl LibraInterface for JsonRpcLibraWrapper {
+    fn libra_timestamp(&self) -> Result<u64, Error> {
+        self.libra_interface.libra_timestamp()
+    }
+
+    fn last_reconfiguration(&self) -> Result<u64, Error> {
+        self.libra_interface.last_reconfiguration()
+    }
+
+    fn retrieve_sequence_number(&self, account: AccountAddress) -> Result<u64, Error> {
+        self.libra_interface.retrieve_sequence_number(account)
+    }
+
+    fn submit_transaction(&self, transaction: Transaction) -> Result<(), Error> {
+        self.submitted_transactions
+            .borrow_mut()
+            .push(transaction.clone());
+        self.libra_interface.submit_transaction(transaction)
+    }
+
+    fn retrieve_validator_config(&self, account: AccountAddress) -> Result<ValidatorConfig, Error> {
+        self.libra_interface.retrieve_validator_config(account)
+    }
+
+    fn retrieve_validator_info(&self, account: AccountAddress) -> Result<ValidatorInfo, Error> {
+        self.libra_interface.retrieve_validator_info(account)
+    }
+}
+
+impl LibraInterfaceTestHarness for JsonRpcLibraWrapper {
+    fn retrieve_discovery_set(&self) -> Result<DiscoverySet, Error> {
+        let account = account_config::discovery_set_address();
+        let account_state = self.libra_interface.get_account_state(account)?;
+        Ok(account_state
+            .get_discovery_set_resource()?
+            .ok_or_else(|| Error::DataDoesNotExist("DiscoverySetResource".into()))?
+            .discovery_set()
+            .clone())
+    }
+
+    fn retrieve_libra_block_resource(&self) -> Result<LibraBlockResource, Error> {
+        let account = account_config::association_address();
+        let account_state = self.libra_interface.get_account_state(account)?;
+        account_state
+            .get_libra_block_resource()?
+            .ok_or_else(|| Error::DataDoesNotExist("BlockMetadata".into()))
+    }
+
+    fn take_all_transactions(&self) -> Vec<Transaction> {
+        self.submitted_transactions
+            .borrow_mut()
+            .drain(0..)
+            .collect()
+    }
+}
+
+// Creates and returns a test node that uses the JsonRpcLibraInterface.
+// This setup is useful for testing nodes as they operate in a production environment.
+fn setup_node_using_json_rpc(config: &NodeConfig) -> (Node<JsonRpcLibraWrapper>, Runtime) {
```

Comment:
why not make this a function within the impl on Node?

Reply:
The impl is generic, whereas these functions are specific to types.. I'm going to leave this as-is for now!